Issue-37 Acceptance tests

phillbarber opened the pull request:
This pull request splits the project into multiple modules.

Annotations - this is so that people's tests can depend on a very small project without bringing in dropwizard.
dummy-tests - Tests that are used to test atam4j (e.g. PassingTest).
acceptance-tests - I looked into incorporating these tests into the main project, however I ran into difficulties with slf4j.  I concluded that it would be best to just keep these tests separate.  The issue was that we'd need to exclude logback-classic from the main atam4j artefact, but include it for tests runs.  It gets more complicated when you consider we are using slf4j-test for some tests but would be using logbackclassic for others.

anuragkapur commented:
@phillbarber - could you please resolve the merge conflicts as reported by the automated check above? Thanks!

phillbarber commented:
Done!

phillbarber commented:
OK, we have some issues with the tests.  It's a timing issue.  

I will work on getting that fixed before we merge this.

...Works on my machine! :)

phillbarber commented:
Have addressed the timing issues in the tests with (hopefully) a nicer solution.  Let me know of any thoughts.

Cheers

anuragkapur commented:
The last build on travis failed with the following error:

```
Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 4.225 sec <<< FAILURE!
givenSampleApplicationStartedWithPassingTest_whenHealthCheckCalledBeforeTestRun_thenTooEarlyMessageReceived(me.atam.atam4jsampleapp.PassingTestAcceptanceTest)  Time elapsed: 3.604 sec  <<< FAILURE!
java.lang.AssertionError: 
Expected: is "Too early to tell - tests not complete yet"
     but: was "All is A OK!"
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.junit.Assert.assertThat(Assert.java:923)
	at me.atam.atam4jsampleapp.AcceptanceTest.checkResponseIsOKAndWithMessage(AcceptanceTest.java:32)
	at me.atam.atam4jsampleapp.PassingTestAcceptanceTest.givenSampleApplicationStartedWithPassingTest_whenHealthCheckCalledBeforeTestRun_thenTooEarlyMessageReceived(PassingTestAcceptanceTest.java:19)
```

Suspect there is an intermittent timing issue somewhere. Thoughts?

anuragkapur commented:
@phillbarber - We should update the "Developing atam4j" section of the README.md with a brief introduction to highlight the purpose of each maven module.

anuragkapur commented on acceptance-tests/src/main/java/me/atam/atam4jsampleapp/ApplicationConfiguration.java:
```diff
@@ -0,0 +1,21 @@
+public class ApplicationConfiguration extends Configuration {
+
+    @JsonProperty
+    private int initialDelayInMillis;
```
`initialDelayInMillis` is specified here but not defined in the corresponding `atam4j/acceptance-tests/src/main/resources/app-config.yml` which is a little confusing. Did you forget something here @phillbarber ?

phillbarber replied:
Didn't forget anything but the current situation is a bit confusing.  I will include this in the yaml file as it makes it clearer.  The variable in the yaml file will represent the default value.  This value is actually set by each test using ConfigOverride objects passed to the DropwizardTestSupport constructor.

anuragkapur commented on acceptance-tests/src/main/java/me/atam/atam4jsampleapp/Atam4jTestApplication.java:
```diff
@@ -0,0 +1,41 @@
+
+public class Atam4jTestApplication extends Application<ApplicationConfiguration> {
+
+    public static final int INITIAL_DELAY = 1000;
```
Unused variable... remove if we don't need it?

anuragkapur commented:
Looking at the number of styling based commits I have made, I think I should add checkstyle which breaks the build if inconsistent styling is used ;)

anuragkapur commented on acceptance-tests/src/test/java/me/atam/atam4jsampleapp/Atam4jApplicationStarter.java:
```diff
@@ -0,0 +1,22 @@
+        DropwizardTestSupport<ApplicationConfiguration> applicationConfigurationDropwizardTestSupport =
+                new DropwizardTestSupport<>(Atam4jTestApplication.class, ResourceHelpers.resourceFilePath("app-config.yml"),
+                        ConfigOverride.config("testClasses", testClass.getName()),
+                        ConfigOverride.config("initialDelayInMillis", initialDelayInMillis + ""));
```
Good practice to use `String.valueOf` to convert int to String :)

anuragkapur commented on acceptance-tests/src/test/java/me/atam/atam4jsampleapp/FailingTestAcceptanceTest.java:
```diff
@@ -0,0 +1,14 @@
+    @Test
+    public void givenSampleApplicationStartedWithFailingTest_whenHealthCheckCalledBeforeTestRun_thenFailureMessageReceived(){
+        applicationConfigurationDropwizardTestSupport = Atam4jApplicationStarter.startApplicationWith(FailingTest.class);
+        checkResponseIsErrorAndWithMessage(AcceptanceTestsHealthCheck.FAILURE_MESSAGE + " 1. Was expecting false to be true", getResponseFromHealthCheck());
```
I missed this in a previous PR, but don't like that use of public attributes in `AcceptanceTestsHealthCheck` (example: FAILURE_MESSAGE). 

If these are to be reused, they should be moved elsewhere. Yes, I can't think of of a possible drawback of using this, but I would vote for avoiding it for the sake of consistency atleast.

Just feels unnatural to use public attributes from a class in a different class. Thoughts? @phillbarber

anuragkapur commented:
I feel the [Favour composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance) principle has been sacrificed across  in the following classes:

* `AcceptanceTest.java`
* `FailingTestAcceptanceTest.java`
* `PassingTestAcceptanceTest.java`

May I suggest a reconsideration of the approach here and possible refactoring @phillbarber ?

anuragkapur commented on atam4j/src/main/java/me/atam/atam4j/Atam4j.java:
```diff
@@ -43,6 +44,9 @@ public Atam4jBuilder(HealthCheckRegistry healthCheckRegistry) {
             this.healthCheckRegistry = Preconditions.checkNotNull(healthCheckRegistry);
         }
 
+        public Atam4jBuilder() {
```
Have removed the unused constructor

anuragkapur commented:
Hey @phillbarber - have added a few new commits and some comments. Sorry, they may only `look` a lot, but all minor comments to gather your thoughts on some coding standards etc. Overall, the refactor looks really good and valuable. Acceptance tests being part of the core build will the library remains solid and error free! Nice work!

phillbarber commented:
Hi Anurag, I think I have addressed each of your points with my commits.  I removed some logic from the base AcceptanceTest class which I think improves things.  This could be taken further (i.e. removing the base class all together) however I quite like the current setup with the @After annotation only declared once so it's not duplicated and possibly forgotten..

Could you send me your code style so I can apply it to my IDE?

anuragkapur commented:
Thanks @phillbarber .. looks good. Merging.

With regards to code style, I don't really have anything automated setup, just tend to refer to something like: https://google.github.io/styleguide/javaguide.html 

However, we can configure checkstyle plugin.. i'll look into it..
